scalyr_agent/copying_manager.py

```diff
@@ -1038,39 +1075,56 @@ def __scan_for_pending_log_files( self ):
 
         # make a shallow copy of pending log_matchers, and pending reloads
         log_matchers = []
-        pending_reload = []
+        pending_reload = {}
         self.__lock.acquire()
         try:
             log_matchers = self.__pending_log_matchers[:]
 
             # get any logs that need reloading and reset the pending reload list
-            pending_reload = self.__logs_pending_reload
-            self.__logs_pending_reload = []
+            pending_reload = self.__logs_pending_reload.copy()
+            self.__logs_pending_reload = {}
         finally:
             self.__lock.release()
 
+        # add new matchers
+        for matcher in log_matchers:
+            self.__dynamic_matchers[matcher.log_path] = matcher
+
         checkpoints={}
 
-        # remove any log processors from the pending_reload list, so they get recreated
-        for matcher in pending_reload:
-            log.log( scalyr_logging.DEBUG_LEVEL_1, "Pending reload for %s" % matcher.log_path )
-            self.__log_paths_being_processed.pop( matcher.log_path, None )
-            # iterate over full list of processors to see if we need to remove any.
-            # We manually iterate so that we can close the processors if necessary
-            # and also so we can get checkpoints so files aren't copied from the beginning
-            # note:  manually calculate the len each loop iteration as it changes if items
-            # are removed
-            i = 0
-            while i < len( self.__log_processors ):
-                p = self.__log_processors[i]
-                if p.log_path == matcher.log_path:
-                    checkpoints[p.log_path] = p.get_checkpoint()
-                    p.close()
-                    del self.__log_processors[i]
-                else:
-                    i += 1
+        # reload the config of any matchers/processors that need reloading
+        reloaded = []
+        for path, log_config in pending_reload.iteritems():
+            log.log( scalyr_logging.DEBUG_LEVEL_1, "Pending reload for %s" % path )
+
+            # only reload matchers that have been dynamically added
+            matcher = self.__dynamic_matchers.get( path, None )
+            if matcher is None:
+                log.log( scalyr_logging.DEBUG_LEVEL_0, "Log matcher not found for %s" % path )
+                continue
+
+            processor_paths = matcher.update_log_entry_config( log_config )
```

Comment:
I'm not a big fan of this interface for `update_log_entry`.  It expects the caller to do too much -- get checkpoints and remember to close the processors.  What if we changed the interface to this:  Inside `update_log_entry_config`, it closes any processors it has... and removes them from its internal list.  It also gets the checkpoints for all processors it is closing and writes that to a passed in `dict` that's an argument on `update_log_entry_config`.

We then use those checkpoints down below when we create the processors.

That leaves how we clean out `__log_paths_being_processed` and `__log_processors`.  Maybe we have a method that iterates over `__log_processors` and removes any processors that are closed, while also removing them from `_log_paths_being_processed`?  It feels like there must be other places where we need this same idea of "removed closed processors and update `_log_paths_being_processed`.

I'm not in love with my suggestion, but I just dislike how much work is being spread out from the `update_log_entry_config` call.